distributed/deploy/adaptive.py

```diff
@@ -188,6 +190,17 @@ def should_scale_up(self):
             if self.scheduler.unrunnable and not self.scheduler.workers:
                 return True
 
+            if not all(ws.processing for ws in self.scheduler.workers.values()):
+                return False
+
+            tasks_processing = sum(
+                (len(w.processing) for w in self.scheduler.workers.values()))
+
+            num_workers = len(self.scheduler.workers)
+
+            if tasks_processing <= num_workers:
+                return False
```

Comment:
I think that all the task_processing count and checks would be better in `needs_cpu()`. The wrong part of code seems to be there: it just take into account the `scheduler.total_occupancy`, which seems to me like the sum of the time of each processing task. It should take into account the number of tasks remaining as pointed out by @delgadom.

Reply:
ok. if we move it into needs_cpu this resolves the problem. that's how I implemented it as a first cut, and I've been running this on a subclassed version of Adaptive on our cluster and it works beautifully. I spun up 10k 15minute jobs with an adaptive cluster with between 0, 1000 nodes and the cluster scaled up then scaled back down as workers completed.

If we go with an implementation where `recommendations()` suggests scaling down based on `workers_to_close`'s recommendation only, I suppose this could involve a race condition where we scale down idle workers before the scheduler has time to transfer tasks to them. So maybe only placing this logic in `needs_cpu` is the right way to go?

I'll modify the PR to only include a change in needs_cpu. I'll try to come up with a less interactive example/test soon, but won't get to implementing real tests until next week.